### ISSUE TITLE
discussion: Creating Soundemote's version of toolchain

### ISSUE THREAD

**Issue by elanhickler** [RobinSchmidt/RS-MET#286]
I think my first step will just be to use toolchain as is. Then I can begin to customize it and let you know the problems I run into, then you can begin to rework some of the code to include additional flexibility that I need.

![image](https://user-images.githubusercontent.com/9116575/62577318-8bdfb100-b853-11e9-980d-14e77b71f9a2.png)

**Comment by RobinSchmidt** [comment]
maybe i could make thses two members pointers instead of direct objects:
```cpp
  ModulationManager modManager;
  AudioModuleFactory moduleFactory;
```
to avoid excessive data objects and allow sharing between several ToolChain objects. and/or maybe that one could be factored out into a baseclass: `  std::vector<AudioModule*> modules;` ...and then you would only need the baseclass and not the full ToolChain class. we'll see

**Comment by elanhickler** [comment]
I started by subclass ToolChain as SeToolChainModule and ToolChainEditor as SeToolChainEditor.

First problem: I need to register my own modules. The constructor of ToolChain registers a bunch of modules and I cannot avoid that because... I don't know how to avoid ToolChain constructor.

![image](https://user-images.githubusercontent.com/9116575/62581641-c0a53580-b85e-11e9-82e0-51b287696ff3.png)

going to unprotected moduleFactory moduleInfos and clear the registered modules for now.

**Comment by elanhickler** [comment]
I need a "getActiveSlot" function as activeSlot is protected.

**Comment by elanhickler** [comment]
removed almost everything, but can't figure out how the selectors and editor is getting its position:

![image](https://user-images.githubusercontent.com/9116575/62583363-c69e1500-b864-11e9-8a55-d0484eb03d55.png)

**Comment by elanhickler** [comment]
progress
![k6Dfy5T 1](https://user-images.githubusercontent.com/9116575/62612785-009b0580-b8bd-11e9-8756-46a96ecc7774.gif)

**Comment by RobinSchmidt** [comment]
>  I need to register my own modules. The constructor of ToolChain registers a bunch of modules and I cannot avoid that because... I don't know how to avoid ToolChain constructor.

right. you can't. i'm starting to think, it would make sense to factor out a baseclass from which you may derive

> can't figure out how the selectors and editor is getting its position...progress

did  you figure it out?

**Comment by elanhickler** [comment]
i figured it out almost... I think I still need to remove your selectors from appearing.

we need an index move function for draggable tabs.

```cpp
template <typename t> void move(int oldIndex, int newIndex, vector<t>& v)
{
	jassert(oldIndex < v.size());
	jassert(newIndex < v.size());
	std::rotate(v.begin() + oldIndex, v.begin() + oldIndex + 1, v.being() + newIndex + 1);
}
```

**Comment by elanhickler** [comment]
tabs now completely work:

![cs1R5nX 1](https://user-images.githubusercontent.com/9116575/62653663-dc6a1380-b912-11e9-8516-1c31bca5398d.gif)

**Comment by elanhickler** [comment]
I think you have a bug

![image](https://user-images.githubusercontent.com/9116575/62655759-88adf900-b917-11e9-809c-f0f919621182.png)

Did you not anticipate having the 0th module removed while modules still exist in other indexes?

**Comment by elanhickler** [comment]
Why does ToolChain keep track of the active slot? Seems that should be on the editor side. I can't think of a way that the active slot has any audio consequences

**Comment by elanhickler** [comment]
progress... I really need to start mixing the outputs of all the tabs. Maybe you should add a "preocessStereoFrameAccumulate" or something to the AudioModule base class and we can override that for all our modules and we can either use "processStereoFrame" when we don't want to mix the input and "processStereoFrameAccumulate" when we do. What do you want to call it? processStereoFrameMixed? processStereoFrameIncludeInput? processStereoFrameWithInput?

![image](https://user-images.githubusercontent.com/9116575/62800451-48c24f80-ba98-11e9-978a-cc1631db7195.png)

**Comment by elanhickler** [comment]
Progress! PrettyScope is now easy to add to an audio module.

![](https://i.imgur.com/T6RXrrj.gif)

**Comment by RobinSchmidt** [comment]
> Did you not anticipate having the 0th module removed while modules still exist in other indexes?

yep - i definitely need to fix this. how could this go undetected for so long?

**Comment by elanhickler** [comment]
because in toolchain it's impossible to remove the 0th module while another module still exists.

**Comment by RobinSchmidt** [comment]
> Setting module to "none" is not removing.

ah - right - i'm leaving a dummy in the slot - leaving the slot as placeholder

**Comment by RobinSchmidt** [comment]
> The index should go to the next available module instead of -1. Edit: It should go to -1 only if there are no modules.

hmmm...i guess the issue here is that i didn't anticipate a completely empty module list. in my toolchain, there is always at least one module present. ...and you may have a completely empty list

**Comment by elanhickler** [comment]
yeah I wanted to be able to have no modules, which is working for my toolchain. I think this is all I had to do, avoid resizing a nullptr editor:

![image](https://user-images.githubusercontent.com/9116575/63380019-72a62c80-c34a-11e9-8e74-55ebdaa7d7dc.png)

**Comment by RobinSchmidt** [comment]
i just added this special case handling to the end of the function
```cpp
void ToolChain::deleteModule(int index)
{
  ScopedLock scopedLock(*lock);
  jassert(index >= 0 && index < size(modules)); // index out of range
  if(activeSlot == index) 
    activeSlot--;
  sendAudioModuleWillBeDeletedNotification(modules[index], index);
  removeFromModulatorsIfApplicable(modules[index]);
  delete modules[index];
  remove(modules, index);

  // handle case when 0th module was deleted (needs test):
  if(activeSlot == -1 && size(modules) > 0 )
    activeSlot = 0;
  // ...in ToolChain, it's impossible to have a completely empty module list, but subclasses (like 
  // Elan's) may want to allow this, so we need to support it
}
```
...not sure, if this is the best  (or even right) way to do it. i guess, i will find out, when i try with your version - because in mine, i don't get into this situation

**Comment by RobinSchmidt** [comment]
i can't build your version due to missing module "elan_juce_helpers".

**Comment by elanhickler** [comment]
Did you want to have a side bar for modules that you drag into the toolchain? If so you probably should clean up your code by supporting no modules. For my toolchain, I'm going to have a menu that pops up and you click to add a module (so, no dragging from one component to another).

https://docs.juce.com/master/classDragAndDropContainer.html
https://docs.juce.com/master/classDragAndDropTarget.html

If you want to do simple dragging of a component like my tabs:

https://docs.juce.com/master/classComponentDragger.html

My implementation is not how you're supposed to do it. You're supposed to make the draggable component have a `ComponentDragger`. Instead, I made the parent component have a ComponentDragger.

1. On mouse down, get component that was clicked, assign to component dragger.

2. On mouse drag, instead of dragging the actual child component, I drag an `ImageComponent` which I assign the image of child component and make it transparent and the user thinks they are dragging the actual child component, but it's just an image of the component. This seemed to work better and be easier to implement then actually moving the component around with the mouse.

```cpp
// ImageComoponent dragImage declared in parent class
dragImage.setImage(tab->createComponentSnapshot(tab->getLocalBounds()));		
dragImage.setAlpha(0.5);
dragImage.setAlwaysOnTop(true);
dragImage.setBounds(tab->getBounds());
addChildComponent(dragImage);
dragger.startDraggingComponent(&dragImage, event);
```

![8caa19fc2b 1](https://user-images.githubusercontent.com/9116575/63384245-4a6efb80-c353-11e9-8138-0299b99961bf.gif)

3. Using `paintOverChildren` method I draw a black transparent square to grey out the child component that is being dragged or a red square if the drop location would be the same place as the child component being dragged. Or draw the red square inbetween tabs where the drop location will be.

The math to check for drop location was really confusing for me to figure out.

```cpp
auto dragX = dragImage.getBounds().getCentreX();

int absDeltaBuffer = INT_MAX;
for (int i = 0; i < getNumTabs(); ++i)
{
    auto* component = getTabButton(i);

    int startX = component->getX();
    int absDelta = abs(dragX - startX);

    if (absDelta > absDeltaBuffer)
        break;

    absDeltaBuffer = absDelta;
    dragPlacementX = startX;
    dragPlacementIndex = i;        
}
     
// see if tab is being placed beyond all tabs (last spot)
if ( (getTabButton(getNumTabs() - 1)->getBounds().getRight() - dragX) < absDeltaBuffer)
    {
        ++dragPlacementIndex;
        dragPlacementX = getTabButton(getNumTabs() - 1)->getBounds().getRight();
    }

// tab original index and 1 more is the same spot
if (dragPlacementIndex > originalDragIndex)
    --dragPlacementIndex;
```

**Comment by elanhickler** [comment]
Robin, you are not following Git repository 101 etiquette!

https://gitlab.com/Hickler/elanjucehelpers

![image](https://user-images.githubusercontent.com/9116575/63385064-13014e80-c355-11e9-9d85-61b02fd4ebb0.png)

**Comment by elanhickler** [comment]
just want to bring your attention to the new ListenerList class in JUCE which now exclusevily uses std::function which makes things so much easier because you can tell your listeners exactly what happened and what function to use, you can give the function a clear name, a lot less confusing. Here's how I use it for my TabbedButtonBar component (children are called 

```cpp
class EditorTabBar : public TabbedButtonBar
{
public:
    EditorTabBar() : TabbedButtonBar(TabbedButtonBar::Orientation::TabsAtTop)
    {
        setInterceptsMouseClicks(true, true);
    }

    class Listener
    {
    public:
        virtual ~Listener() = default;

        virtual void tabWasRemoved(int index) {}
        virtual void tabWasAdded(int index) {}
        virtual void tabWasMoved(int oldIndex, int newIndex) {}
        virtual void tabWasActivated(int index) {}
    };

    void addListener(Listener* l) { listeners.add(l); }
    void removeListener(Listener* l) { listeners.remove(l); }

    virtual void moveTab(const int currentIndex, const int newIndex, const bool animate, bool notifyListeners = true, EditorTabBar::Listener* callingObject = nullptr)
    {
        TabbedButtonBar::moveTab(currentIndex, newIndex, animate);

        if (currentIndex != newIndex)
            if (notifyListeners)
                listeners.callExcluding(callingObject, [&](Listener& listener) { listener.tabWasMoved(currentIndex, newIndex); });
    }

   
    virtual void addTab(const String& tabName, Colour tabBackgroundColour, int insertIndex, bool notifyListeners = true, EditorTabBar::Listener* callingObject = nullptr)
    {
        TabbedButtonBar::addTab(tabName, tabBackgroundColour, insertIndex);

        if (notifyListeners)
            listeners.callExcluding(callingObject, [&](Listener& listener) { listener.tabWasAdded(insertIndex == -1 ? getNumTabs() - 1 : insertIndex); });
    }

     // etc...
protected:
    ListenerList<Listener> listeners;
```

**Comment by RobinSchmidt** [comment]
> read the readme

ah - ok - i see. but it seems to have problems with using the functionality from there. i'm getting a bunch of compile errors of the type:
```
BasicEditor.h(87,1): error C2653:  'FileHelper': is not a class or namespace name
BasicEditor.h(87,34): error C3861:  'toUniversalSlash': identifier not found
```
i can see the class and function - but apparently, the compiler can't. is some include messed up? or a namespace issue? dunno

**Comment by elanhickler** [comment]
you're compiling Soundemote\AudioPlugins\SeToolChain ?

I'm going to re-make the project with jucer now and try.

**Comment by elanhickler** [comment]
there might be some include issues, pull the latest from ElanJuceHelpers, just updated it.

**Comment by RobinSchmidt** [comment]
ok - i can compile now - but i had to add the /bigobj option to the compiler settings (of the "...SharedCode" project):
![image](https://user-images.githubusercontent.com/1833099/63588897-e4c77d00-c5a7-11e9-84ce-eac459eded2e.png)
without it, i got errors that said something about object files exceeding maximum allowed size (in debug/x64 mode)

**Comment by RobinSchmidt** [comment]
> still trying to release the new Flower Child Filter, have you solved all the build errors you can?

didn't you say in you mail that you have fixed the build issues? after updating my codeblocks, i got some new errors which i'm still struggling with. problem is, even the new codeblocks (including a new version of mingw) doesn't allow x64 builds - and juce doesn't seem to support being compiled in x86 anymore - wehn trying, i get compiler error from juce. hmm...maybe i should try on linux

**Comment by RobinSchmidt** [comment]
okay - i *may* have fixed the crash. drag an update of my library  (code changes only affect my library). you'll find a new comment in ToolChain's destructor about what - i think - may have gone wrong

**Comment by elanhickler** [comment]
in jura_AudioPlugin.cpp I added what is in the red square to prevent crash. For some reason all of a sudden this started happening, also keep in mind the getXmlFromBinary may return nullptr according to the description, so it's probably a good idea to have this in the first place:

![image](https://user-images.githubusercontent.com/9116575/63638986-851eaf80-c643-11e9-847d-7c4b7e64ccbd.png)

**Comment by elanhickler** [comment]
in jura_ModulatableParameter.cpp getting crash if I add TorusGenerator to my SeToolChain. The crash happens if you add it then remove it before closing SeToolChain. The crash does not happen if you never load TorusGenerator.

![image](https://user-images.githubusercontent.com/9116575/63639027-f0688180-c643-11e9-8841-8b02e2e602a3.png)

**Comment by elanhickler** [comment]
Crash if I register Flower Child Filter:

![image](https://user-images.githubusercontent.com/9116575/63639161-222e1800-c645-11e9-995c-8c0c08736be4.png)

![image](https://user-images.githubusercontent.com/9116575/63639168-34a85180-c645-11e9-8a8e-c82b3ff1f915.png)
